File: scala/play/book-api/app/repositories/BookRepository.scala
package repositories

import scala.concurrent.Future
import entities.Book

import java.util.UUID

trait BookRepository {
  def findAll(): Future[List[Book]]
  def findById(id: UUID): Future[Option[Book]]
  def insert(book: Book): Future[Either[Error, Book]]
  def update(id: UUID, book: Book): Future[Int]
  def delete(id: UUID): Future[Int]
}

File: go/gin/book-api/repositories/book_repository.go
package repositories

import (
	"github.com/google/uuid"
	"github.com/guisofiati/book-api-frameworks/go/gin/book-api/models"
)

type BookRepository interface {
	FindAll() []models.Book
	FindById(id uuid.UUID) (*models.Book, error)
	Insert(entity models.Book) *models.Book
	Update(id uuid.UUID, entity models.Book) (*models.Book, error)
	Delete(id uuid.UUID) error
}
